src/wazuh_db/wdb_global.c

```diff
@@ -1415,6 +1415,251 @@ cJSON* wdb_global_get_agents_by_connection_status (wdb_t *wdb, int last_agent_id
+        os_realloc(backups_time_array, number_of_files * sizeof(time_t), backups_time_array);
+
+        snprintf(tmp_path, OS_SIZE_512, "%s/%s", WDB_BACKUP_FOLDER, entry->d_name);
+        if(!stat(tmp_path, &backup_info)) {
+            os_strdup(tmp_path, backups_array[number_of_files-1]);
+            backups_time_array[number_of_files-1] = current_time - backup_info.st_mtime;
+        } else {
+            backups_array[number_of_files-1] = NULL;
+            backups_time_array[number_of_files-1] = -1;
+        }
+    }
+
+    backups_to_delete = number_of_files - wconfig.wdb_backup_settings[0].max_files;
+
+    for(int i = 0; backups_to_delete > i; i++) {
+        int oldest_backup = -1;
+        int array_index = 0;
+        char* backup_to_remove = NULL;
+        for(int j = 0; j < number_of_files; j++ ){
+            if(backups_array[j] && backups_time_array[j] > oldest_backup) {
+                backup_to_remove = backups_array[j];
+                oldest_backup = backups_time_array[j];
+                array_index = j;
+            }
+        }
+
+        if(backup_to_remove) {
+            unlink(backup_to_remove);
+            backups_time_array[array_index] = -1;
+        }
+    }
+
+    for(int j = 0; j < number_of_files; j++) {
+        os_free(backups_array[j]);
+    }
+
+    os_free(backups_array);
+    os_free(backups_time_array);
+    closedir(dp);
+
+    return OS_SUCCESS;
+}
+
+cJSON* wdb_global_get_backup() {
+    DIR* dp = NULL;
+    cJSON* j_backups = NULL;
+    struct dirent *entry = NULL;
+
+    dp = opendir(WDB_BACKUP_FOLDER);
+
+    if(!dp) {
+        mdebug1("Unable to open backup directory '%s'", WDB_BACKUP_FOLDER);
+        return NULL;
+    }
+
+    j_backups = cJSON_CreateArray();
+    while (entry = readdir(dp), entry) {
+        if ((strcmp(entry->d_name, ".") == 0) || (strcmp(entry->d_name, "..") == 0)) {
+            continue;
+        }
+
+        cJSON_AddItemToArray(j_backups, cJSON_CreateString(entry->d_name));
+    }
+
+    closedir(dp);
+    return j_backups;
+}
+
+int wdb_global_restore_backup(wdb_t** wdb, char* snapshot, bool save_pre_restore_state, char* output) {
+    bool backup_found = false;
+    DIR* dp = NULL;
+    struct dirent *entry = NULL;
+    struct stat backup_info = {0};
+    time_t last_modification_time = -1;
+    char* most_recent_backup = NULL;
+    char global_path[OS_SIZE_256] = {0};
+    char global_tmp_path[OS_SIZE_256] = {0};
+    char global_pre_restore_path[OS_SIZE_256] = {0};
+    char backup_to_restore_path[OS_SIZE_256] = {0};
+    char tmp_path[OS_SIZE_512] = {0};
+
+    dp = opendir(WDB_BACKUP_FOLDER);
+
+    if(!dp) {
+        mdebug1("Unable to open backup directory '%s'", WDB_BACKUP_FOLDER);
+        return OS_INVALID;
+    }
+
+    while (entry = readdir(dp), entry) {
+        if ((strcmp(entry->d_name, ".") == 0) || (strcmp(entry->d_name, "..") == 0)) {
+            continue;
+        }
+
+        if(snapshot) {
```

Comment:
I think we can leave this logic outside the while loop.
If a specific snapshot is specified we should try to open it, otherwise, we should search for the newest one.

Reply:
Fixed!